tests/integrations/gcp/test_gcp.py

```diff
@@ -265,3 +268,79 @@ def cloud_function(functionhandler, event):
     assert exception["type"] == "Exception"
     assert exception["value"] == "something went wrong"
     assert exception["mechanism"] == {"type": "gcp", "handled": False}
+
+
+def test_traces_sampler_gets_correct_values_in_sampling_context(
+    run_cloud_function, DictionaryContaining  # noqa:N803
+):
+    import inspect
+
+    envelope, event, stream_data = run_cloud_function(
+        dedent(
+            """
+            functionhandler = None
+            event = {
+                "type": "chase",
+                "chasers": ["Maisey", "Charlie"],
+                "num_squirrels": 2,
+            }
+            def cloud_function(functionhandler, event):
+                # this runs after the transaction has started, which means we
+                # can make assertions about traces_sampler
+                try:
+                    traces_sampler.assert_any_call(
+                        DictionaryContaining({
+                            "gcp_env": DictionaryContaining({
+                                "function_name": "chase_into_tree",
+                                "function_region": "dogpark",
+                                "function_project": "SquirrelChasing",
+                            }),
+                            "gcp_event": {
+                                "type": "chase",
+                                "chasers": ["Maisey", "Charlie"],
+                                "num_squirrels": 2,
+                            },
+                        })
+                    )
+                except AssertionError as e:
+                    # catch the error and print it because the error itself will
+                    # get swallowed by the SDK as an "internal exception"
+                    print("\\nAssertionError: {}\\n".format(e))
+
+                return "dogs are great"
+            """
+        )
+        + FUNCTIONS_PRELUDE
+        + dedent(inspect.getsource(DictionaryContaining))
```

Comment:
can we not return the traces sampler calls as response body and do assertions on teh response body? `assert x not in` can easily pass if it shouldn't, as the log buffer has a limited size the assertionerrors can be flushed out.


```
calls = []
def traces_sampler(...):
    calls.append(...)
    return ...

init(traces_sampler=traces_sampler)


def cloud_function(...):
    return json.dumps(calls)

Reply:
Yeah, I thought about the flushing problem, and wondered if maybe I should add an envelope processor which deletes all of the envelope's data, so that there'd be very little printed after the assertion error, which would (more or less) guarantee the error wouldn’t scroll off the top of the buffer.

For GCS the above should work fine (though in GCS we're running it locally, so there isn't a limited buffer). For AWS, the one tricky part is that I'm comparing to a class (`LambdaContext`) which only exists in the AWS runtime AFAICT. But your point about printing is valid. I didn't originally have these assertions in the function handler itself, but now that I do, it occurs to me I can just return the error (if it happens).